dbt/include/snowflake/macros/relations/table/create.sql

```diff
@@ -46,7 +46,8 @@
       {%- endif -%}
 
   {%- elif language == 'python' -%}
-    {{ py_write_table(compiled_code=compiled_code, target_relation=relation, temporary=temporary) }}
+    {%- set table_type = 'transient' if transient else '' -%}
+    {{ py_write_table(compiled_code=compiled_code, target_relation=relation, temporary=temporary, table_type=table_type) }}
```

Comment:
Change
```sql
{%- set table_type = 'transient' if transient else '' -%}
{{ py_write_table(compiled_code=compiled_code, target_relation=relation, temporary=temporary, table_type=table_type) }}
```

to
```sql
{{ py_write_table(compiled_code=compiled_code, target_relation=relation, temporary=temporary, transient=transient) }}
```

and handle within `py_write_table` instead if we want to proceed with Mike's suggestion above (https://github.com/dbt-labs/dbt-snowflake/pull/777/files#r1330527575).

Reply:
I would do a few things here in alignment with my comment above. I would update the definition of `table_type` here to consider `temporary` as discussed above. And then I would not pass `temporary` into the `py_write_table` call at all. That tidies up all the config parsing. You would still need to keep that if block in the `py_write_table` macro in the even that other folks are using it; but then when we eventually retire the `temporary` argument in that macro, we don't need to remember to come back here and deal with it here.

Reply:
@dbeatty10 I think this still aligns with what you're saying above, correct? We're effectively forcing the use of `table_type` here.

Reply:
@mikealfare there's one more thing to handle here:
- 👉 dbt-snowflake has a boolean config named [`transient `](https://docs.getdbt.com/reference/resource-configs/snowflake-configs#configuring-transient-tables-in-dbt_projectyml) that [defaults to `None`](https://github.com/dbt-labs/dbt-snowflake/blob/273694b38ed05d046a6b284a3ac396478a241bfa/dbt/adapters/snowflake/impl.py#L25)

### Which brings us to three-valued logic ...

There's three-valued logic to handle for the `transient` boolean config:
- `None` (pass-through and let `py_write_table` decide what to do)
- `True` (use `transient`)
- `False` (use `permanent`?)

### How 'bout this?

So I think we'd need handle the null case first to be sure the `table_type` is set (or not set!) correctly:

```sql
{% if transient is none %}
    {% set table_type = none %}
{% elif transient %}
    {% set table_type = "transient" %}
{% else %}
    {% set table_type = "permanent" %}
{% endif %}

{{ py_write_table(compiled_code=compiled_code, target_relation=relation, table_type=table_type) }}
```

This is assuming an implementation of `py_write_table` like [our most recent iteration](https://github.com/dbt-labs/dbt-snowflake/pull/777#discussion_r1353680684).

Reply:
If `transient` is `None`, then wouldn't we set `table_table` based on `temporary` since we're not passing `temporary` into `py_write_table` anymore? Put another way, I don't think you would be able to create a table with `table_type="temporary"` in your logic flow without also passing `temporary` into the call to `py_write_table`. And I think we don't want to do that from what we said.

```
{% if transient is none and temporary %}
    {% set table_type = "temporary" %}
{% elif transient is none %}
    {% set table_type = none %}
{% elif transient %}
    {% set table_type = "transient" %}
{% else %}
    {% set table_type = "permanent" %}
{% endif %}
```

Reply:
So we can assess all the proposed logic with full context, I'm going to raise an alternative PR that is branched off of this one.

Should be able to do that sometime within the next 2 hours.

Then we can hammer on it to make sure all the corner cases are covered.

Reply:
@mikealfare and @jeremyyeo

[Here](https://github.com/dbt-labs/dbt-snowflake/compare/0bf3822d90d7c64b15270e5e510861786562a92c...7074d92cebc2f5d8b435b5bf2f88e9d260f6cdfe) is some **untested**(!) draft code that I think is backwards-compatible if anyone is relying on the original signature:
```sql
py_write_table(compiled_code, target_relation, temporary=False)
```

That draft code can be simplified quite a bit if we choose to make an breaking/incompatible change of the signature to be this instead:
```sql
py_write_table(compiled_code, target_relation, table_type="")
```

Reply:
The `table.sql` file in your draft code looks right. But the `create.sql` file doesn't:

```
{% if temporary -%}
  {%- set table_type = "temporary"
{%- elif transient -%}
  {%- set table_type = "transient"
{%- else -%}
  {%- set table_type = ""
{%- endif %}
```

If I'm following, we said that we want the use of `table_type`, which I think aligns with the use of `transient`, to override `temporary`. However, in the code block it looks like `temporary` takes effect over `transient`. Is that right?

Reply:
Yes, sharp eyes -- you read that right!

After examining the [logic for `sql` tables](https://github.com/dbt-labs/dbt-snowflake/blob/ff08a1cd9ea61bb5e54441a4a84d507ca461eedd/dbt/include/snowflake/macros/relations/table/create.sql#L20-L24) in dbt-snowflake it became clear that it prioritizes the `temporary` config over the `transient` config.

So I flipped-flopped from our earlier discussions and switched the implementation to standardize on the behavior of `sql` tables.

Specifically, I just re-factored the code so that [this logic](https://github.com/dbt-labs/dbt-snowflake/blob/ff08a1cd9ea61bb5e54441a4a84d507ca461eedd/dbt/include/snowflake/macros/relations/table/create.sql#L20-L24) applies to **both** `sql` and `python` tables.

Here is the relevant portion of the diff:

<img width="1314" alt="image" src="https://github.com/dbt-labs/dbt-snowflake/assets/44704949/3f13f6d9-caf5-4849-a924-522abc3f6d03">

Reply:
@mikealfare Currently failing CI, but draft PR is up here: https://github.com/dbt-labs/dbt-snowflake/pull/802

Reply:
I put a comment there. You're missing the close braces on the sets, so the code quality failed. Will we be moving forward with 802 then as the primary PR?

Reply:
Yes, let's consider https://github.com/dbt-labs/dbt-snowflake/pull/802 the primary PR. It preserves @jeremyyeo's commits history and authorship credit.